### PR TITLE
Replace QUUID with boost::uuids::uuid

### PR DESCRIPTION
**Pull request by lefticus**
@kbenne note: aparently boost::uuid has two problems

  1. Abhorent speed if you recreate the generator on each call
  2. default constructor makes no guarantee on value, nil generator
     needs to be used instead

We fix those issues by making our own UUID type that inherits
from boost::uuids::uuid and sets the default construction.

We also make global thread safe generators that can be called as
necessary. That code can be moved to C++11's thread_local at some point
in the future if you move to MSVC2015 as a requirement.

Important notes:

  * The built in QUUID::toString was leading to much duplicated and
    inefficient string/qstring/uuid conversions. This patch reduces that
  * We rename methods as necessary to get it looking as much like QUUID as
    possible.

### CODE REVIEW

**Comment by kbenne** [comment]
Thank you @lefticus. We are wrapping up the 1.11.0 release right now, but as soon as that is out I want to merge this before it gets stale.

File: openstudiocore/src/utilities/core/UUID.i [hunk 1]
```diff
@@ -12,16 +12,13 @@
 namespace openstudio{
 
   #ifdef SWIGRUBY
-    %rename("nil?") isNull();
+    %rename("nil?") is_nil();
```
Comment (macumber) [inline openstudiocore/src/utilities/core/UUID.i]: Does this work?

File: openstudiocore/src/utilities/core/UUID.i [hunk 2]
```diff
@@ -12,16 +12,13 @@
...
   #endif
 
   class UUID {
   public:
-    bool operator!= ( const UUID & other ) const;
-    bool operator< ( const UUID & other ) const;
-    bool operator== ( const UUID & other ) const;
-    bool operator> ( const UUID & other ) const;
-    bool isNull () const;
+    bool is_nil() const;
```
Comment (macumber) [inline openstudiocore/src/utilities/core/UUID.i]: I'd prefer we only export one of these, let's just stick with isNull

File: openstudiocore/src/utilities/core/UUID.i (hunk 1 above)
Comment (lefticus) [inline openstudiocore/src/utilities/core/UUID.i]: I cannot imagine why it wouldn't work? It's been that way for eons AFAICT. All I did was change the name to match the new uuid class.

Comment (lefticus) [inline openstudiocore/src/utilities/core/UUID.i]: Plus, I of course verified that all existing ruby uuid related unit tests still pass.

Comment (macumber) [inline openstudiocore/src/utilities/core/UUID.i]: I was originally confused thinking:

    if !model_object.handle

would work, but I forgot that Ruby doesn't implicitly call `.nil?` in that case.  As for the name, I've never called `.nil?`, I always use `isNull`, so we should keep all of those.  What is the benefit of also adding `is_nil`?  Is there any reason to add a third method for that?

Comment (lefticus) [inline openstudiocore/src/utilities/core/UUID.i]: `is_nil` is the native name for the `boost::uuids::uuid` "check if null" method. So I added it just for completeness. We can probably safely remove it.

The rest of the API remains unchanged from the ruby perspective.

**Comment by nrel-bot-2** [comment]
@kbenne @lefticus @lgentile it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@kbenne @lefticus @lgentile it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@kbenne @lefticus @lgentile it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@kbenne @lefticus @lgentile it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@kbenne @lefticus @lgentile it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@kbenne @lefticus @lgentile it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@kbenne @lefticus @lgentile it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@kbenne @lefticus @lgentile it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@kbenne @lefticus @lgentile it has been 7 days since this pull request was last updated.

**Pull request closed without merging**